Add basic tests for RStudio

dcermak opened the pull request:
- Related ticket: N.A.
- Needles: ~~https://github.com/os-autoinst/os-autoinst-needles-opensuse/pull/634~~, https://github.com/os-autoinst/os-autoinst-needles-opensuse/pull/639
- Verification run: ~~https://openqa.opensuse.org/t1173227~~ ~~https://openqa.opensuse.org/tests/1173265~~

~~Unfortunately, I don't really know how to make a verification run with this, as it requires additional needles that are not present on o3.~~ Also, it requires the creation of a new test suite with the following settings:
```
BOOT_HDD_IMAGE=1
DESKTOP=gnome
HDD_1=%DISTRI%-%VERSION%-%ARCH%-%BUILD%-%DESKTOP%-x11@%MACHINE%.qcow2
NOAUTOLOGIN=0
REGRESSION=gnome
START_AFTER_TEST=create_hdd_gnome-x11
UEFI_PFLASH_VARS=%DISTRI%-%VERSION%-%ARCH%-%BUILD%-gnome-x11@%MACHINE%-uefi-vars.qcow2
YAML_SCHEDULE=schedule/rstudio.yaml
```

okurz commented on lib/rstudio.pm:
```diff
@@ -0,0 +1,111 @@
+        if (check_screen('generic-desktop')) {
+            last;
+        }
+    }
```
There is `send_key_until_needlematch` which you can use to replace the complete loop.

dcermak replied:
Didn't know that function, thank you!

okurz commented on tests/x11/rstudio_desktop.pm:
```diff
@@ -0,0 +1,41 @@
+    ensure_installed('rstudio-desktop');
+
+    x11_start_program('rstudio', target_match => 'rstudio_desktop-main-window');
+    wait_still_screen;
```
this invests at least 7 seconds. Is this necessary?

okurz commented on tests/x11/rstudio_desktop.pm:
```diff
@@ -0,0 +1,41 @@
+    rstudio_create_and_test_new_project(rstudio_mode => "desktop");
+
+    # bye-bye
+    wait_screen_change { send_key('alt-f4') };
```
You should not need a `wait_screen_change` as for all "x11" tests the post_run_hook ensures to wait for "generic-desktop". You overwrite the post_run_hook but then I suggest to still call the base class one additionally. E.g. after line 38 add `$self->SUPER::post_run_hook;`. As an alternative replace the `wait_screen_change` with an `assert_screen 'generic-desktop';`

dcermak replied:
I have opted to adding `$self->SUPER::post_run_hook;`

okurz commented on tests/x11/rstudio_prepare.pm:
```diff
@@ -0,0 +1,32 @@
+
+sub run() {
+    # increase timeout to 15 mins, shouldn't take as long, but who knows...
+    # FIXME: remove git
```
so what about the FIXME?

dcermak replied:
I need to submit a fix to rstudio to be able to remove this.

okurz replied:
ok then please avoid the word "FIXME" here. They tend to stay in code forever :)

dcermak replied:
So should I remove it and make a TODO note for myself?

okurz replied:
yes, please. And just in case you never come to it the test will still be valid and work :)

dcermak replied:
And it's gone!

okurz commented on tests/x11/rstudio_server.pm:
```diff
@@ -0,0 +1,67 @@
+    # open R Studio in the browser and don't verify that Firefox actually
+    # started via a needle
+    x11_start_program("firefox http://localhost:8787", valid => 0);
+    wait_still_screen;
```
that's a waste of time. It should be enough in the next step to just look for the needle area that you want to click.

okurz commented on tests/x11/rstudio_server.pm:
```diff
@@ -0,0 +1,67 @@
+    send_key('ret');
+
+    # reject saving the password for bernhard but don't fail in case this doesn't show up
+    check_screen('rstudio_server-firefox-save-password-prompt', timeout => 10) && assert_and_click('rstudio_server-firefox-save-password-prompt', timeout => 1);
```
Please use a multi-tag `assert_screen` with `match_has_tag` instead of `check_screen` with non-zero timeout to prevent introducing any timing dependant behaviour, to save test execution time as well as state more explicitly from the testers point of view what are the expected alternatives. For example:

```perl
assert_screen([qw(yast2_console-finished yast2_missing_package)]);
if (match_has_tag('yast2_missing_package')) {
    send_key 'alt-o';  # confirm package installation
    assert_screen 'yast2_console-finished';
}
```

okurz commented on tests/x11/rstudio_server.pm:
```diff
@@ -0,0 +1,67 @@
+    assert_and_click("rstudio_server-sign-out");
+    assert_screen("rstudio_server-login-username");
+    send_key('alt-f4');
+    check_screen('rstudio_server-firefox_quit-and-close-tabs', timeout => 10) && assert_and_click('rstudio_server-firefox_quit-and-close-tabs', timeout => 1);
```
same as above

ggardet commented:
You should probably set `rstudio_prepare` as milestone. That way, if `rstudio_desktop`, `rstudio_server` will start from a clean state.

dcermak commented:
@okurz I've addressed your comments and re-run this on o3.

@ggardet The ARM tests fails, for some reason the needle that should click on a specific checkbox selects another one?! This really makes no sense, if you compare the step on x86_64: https://openqa.opensuse.org/tests/1173265#step/rstudio_desktop/39 with ARM: https://openqa.opensuse.org/tests/1173228#step/rstudio_desktop/40 you see that these are different needles!

ggardet commented:
> @ggardet The ARM tests fails, for some reason the needle that should click on a specific checkbox selects another one?! This really makes no sense, if you compare the step on x86_64: https://openqa.opensuse.org/tests/1173265#step/rstudio_desktop/39 with ARM: https://openqa.opensuse.org/tests/1173228#step/rstudio_desktop/40 you see that these are different needles!

It does make sense. There are 2 identical checkbox. openQA will match one, or the other! I guess you should match the entire line instead.

dcermak commented:
It now passes on ARM too after fixing the needles: https://openqa.opensuse.org/tests/1173286

okurz commented:
looks great now! :)
